File: Diagrammatize.Design.Pattern/code/cxx/Decorator/A1/UpDownBorder.cpp
#include "UpDownBorder.hpp"

#include <sstream>

UpDownBorder::UpDownBorder(Display* display, char ch): Border(display), borderChar(ch) {
}

int UpDownBorder::getColumns() {
    return display->getColumns();
}

int UpDownBorder::getRows() {
    return 1 + display->getRows() + 1;
}

std::string UpDownBorder::getRowText(int row) {
    if (row == 0 || row == getRows() - 1) {
        return makeLine(borderChar, getColumns());
    } else {
        return display->getRowText(row - 1);
    }
}
    
std::string UpDownBorder::makeLine(char ch, int count) {
    return std::string(count, ch);
}

File: Diagrammatize.Design.Pattern/code/python/Decorator/A1/UpDownBorder.py
from Border import Border

class UpDownBorder(Border):
    def __init__(self, display, ch):
        super().__init__(display)
        self.borderChar = ch

    def getColumns(self):
        return self.display.getColumns()

    def getRows(self):
        return 1 + self.display.getRows() + 1

    def getRowText(self, row):
        if row == 0 or row == self.getRows() - 1:
            return self.makeLine(self.borderChar, self.getColumns())
        else:
            return self.display.getRowText(row-1)

    def makeLine(self, ch, count):
        return str(ch)*count

